src/hotspot/share/utilities/vmError.cpp

```diff
@@ -1312,6 +1313,38 @@ int VMError::prepare_log_file(const char* pattern, const char* default_pattern,
   return fd;
 }
 
+// Attempt to set the state to Native in VMError.
+// By setting it to Native, it allows os::fork_and_exec to execute cmd such as jcmd %p.
+// Otherwise, it may end up with a deadlock when cmd tries to synchronize all Java threads
+// at safepoints.
+class VMErrorThreadToNativeFromVM : public StackObj {
+ private:
+   JavaThread* _thread;
+
+ public:
+  VMErrorThreadToNativeFromVM(Thread* t) : _thread(nullptr) {
+    if (t != nullptr && t->is_Java_thread() &&
+        !Threads_lock->owned_by_self()) { // VMError::controlled_crash() grabs Threads_lock.
+      _thread = JavaThread::cast(t);
+      assert(_thread == Thread::current(), "must be current thread");
+      assert(_thread->thread_state() == _thread_in_vm, "must be in VM");
+    }
+
+    if (_thread != nullptr) {
+      assert(!_thread->owns_locks(), "must release all locks when leaving VM");
+      ThreadStateTransition::transition_from_vm(_thread, _thread_in_native);
+    }
+  }
+
+  ~VMErrorThreadToNativeFromVM() {
+    if (_thread != nullptr) {
+      ThreadStateTransition::transition_from_native(_thread, _thread_in_vm);
+      assert(!_thread->is_pending_jni_exception_check(), "Pending JNI Exception Check");
+      // We don't need to clear_walkable because it will happen automagically when we return to java
```

Comment:
We are not executing JNI code when the do the fork_and_exec so this does not seem necessary.
The comment about `clear_walkable` also doesn't make sense here - we are crashing so we are not returning to Java at all.

Reply:
Also, recursive assertions don't work here.